### PR TITLE
LibGfx: Support JPEG SOF2 images :yakbait:

### PR DESCRIPTION
**Pull request by LucasChollet**
There are two flavor of SOF2 images, spectral selection and successive approximation (which is used in combination with spectral selection).

This PR adds support for SOF2 that use only spectral selection (actually it was already kinda supported, it just needed support for End-of-Bands).

This PR exists in order to receive merge happiness to compensate how successive approximations hurt my brain 🫠.

Battery-included PR, am I becoming addicted to bonus points given for each test added?

### CODE REVIEW

**Comment by Hendiadyoin1** [comment]
Can you check on the copy-right of the image?

**Comment by LucasChollet** [comment]
Oops, I didn't think of it, I will replace it

File: Userland/Libraries/LibGfx/JPEGLoader.cpp [hunk 1]
```diff
@@ -1178,8 +1178,7 @@ static bool is_app_marker(Marker const marker)
 
 static bool is_miscellaneous_or_table_marker(Marker const marker)
 {
-    // B.6 - Summary
-    // See: Figure B.17 – Flow of marker segment
```
Comment (nico) [inline Userland/Libraries/LibGfx/JPEGLoader.cpp:1180]: Does it make sense to also link to the figure? (As-is is fine too, up to you.)

File: Userland/Libraries/LibGfx/JPEGLoader.cpp [hunk 2]
```diff
@@ -313,7 +313,13 @@ static inline i32* get_component(Macroblock& block, unsigned component)
 
 static ErrorOr<void> add_dc(JPEGLoadingContext& context, Macroblock& macroblock, ScanComponent const& scan_component)
 {
-    auto& dc_table = context.dc_tables.find(scan_component.dc_destination_id)->value;
```
Comment (nico) [inline Userland/Libraries/LibGfx/JPEGLoader.cpp:312]: As you're likely aware, you could also do:

```
    auto dc_table_iterator = context.dc_tables.find(scan_component.dc_destination_id);
    if (dc_table_iterator == context.dc_tables.end()) {
        // error
    }
    auto& dc_table = dc_table_iterator->value;
```

Then it's more like the lhs. But as-is is fine too.

File: Userland/Libraries/LibGfx/JPEGLoader.cpp [hunk 3]
```diff
@@ -313,7 +313,13 @@ static inline i32* get_component(Macroblock& block, unsigned component)
...
+    auto maybe_table = context.dc_tables.get(scan_component.dc_destination_id);
+    if (!maybe_table.has_value()) {
+        dbgln_if(JPEG_DEBUG, "Unable to find a DC table with id: {}", scan_component.dc_destination_id);
```
Comment (nico) [inline Userland/Libraries/LibGfx/JPEGLoader.cpp:318]: Is there much value in logging this? You'll see the actual error right after anyways, right? (Up to you.)

File: Userland/Libraries/LibGfx/JPEGLoader.cpp [hunk 4]
```diff
@@ -649,8 +650,7 @@ static ErrorOr<void> read_start_of_scan(AK::SeekableStream& stream, JPEGLoadingC
         current_scan.spectral_selection_end,
         current_scan.successive_approximation);
 
-    // The three values should be fixed for baseline JPEGs utilizing sequential DCT.
-    if (current_scan.spectral_selection_start != 0 || current_scan.spectral_selection_end != 63 || current_scan.successive_approximation != 0) {
+    if (current_scan.spectral_selection_start > 63 || current_scan.spectral_selection_end > 63 || current_scan.successive_approximation != 0) {
```
Comment (nico) [inline Userland/Libraries/LibGfx/JPEGLoader.cpp:653]: Should this get a `// FIXME: Support progressive jpegs with current_scan.successive_approximation != 0`?

**Review by nico — APPROVED** [review]
Very cool! LGTM. (Any reason this is a draft?)

All comments below are nits and optional.

**Comment by LucasChollet** [comment]
Thanks! The reason why this is a draft is that the test image that I generate to replace the copyrighted image doesn't work :upside_down_face:. AFAICT, it's unrelated, but I'm trying to solve it first.

File: Userland/Libraries/LibGfx/JPEGLoader.cpp (hunk 2 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/JPEGLoader.cpp:312]: As `dc_table_iterator == context.dc_tables.end()` is what `get` do. I think that it's what we should call

File: Userland/Libraries/LibGfx/JPEGLoader.cpp (hunk 3 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/JPEGLoader.cpp:318]: The message gives you the table ID and the error doesn't.
That's not super-duper useful, but it won't hurt either

**Comment by LucasChollet** [comment]
Since nico's review, I added the last commit.

The bug I face is actually 2c98eff not being a good enough hack and not working for some images dimension. So I guess this PR can be merged as the bug is unrelated. However, some SOF2 image won't work, even if this PR advertises it.

**Comment by nico** [comment]
I think merging this as-is is fine as it's a definite improvement about the current state :) So I think you can mark this as not-a-draft.

**Comment by nico** [comment]
Very beautiful artwork in the replacement test image btw!

**Comment by LucasChollet** [comment]
The issue is that the bug provoke a crash, so as an example, from fuzzer point of view it worsens the decoder.
The good new is that I found a solution right now :smile:

**Comment by LucasChollet** [comment]
Last push move the G.1.2.2 section after A.2.3 in `build_macroblocks`, commit [LibGfx: Support End-Of-Bands run](https://github.com/SerenityOS/serenity/pull/17652/commits/5802a2bcf0764a870fe514e9d61cef46ce8f6e4a).
It has no impact here but ease the introduction of #17709

**Review by nico — APPROVED** [review]
Still lg

**Review by linusg — CHANGES_REQUESTED** [review]
Please re-read all the commit message titles and make sure it's clear they're talking about JPEGs. LibGfx is massive. `LibGfx: Support End-Of-Bands run` ain't it :)

Also `magic variables` -> `magic numbers` (https://en.wikipedia.org/wiki/Magic_number_(programming))

**Review by nico — APPROVED** [review]
i can't ui

**Pull request merged**